sidecar/dns/server.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		logrus.WithError(err).Errorf("Error writing DNS response")
+	}
+}
+
+func (s *Server) handleQuestion(question dns.Question, request, response *dns.Msg) error {
+
+	switch question.Qclass {
+	case dns.ClassINET:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question class: %v", dns.Class(question.Qclass).String())
+	}
+
+	switch question.Qtype {
+	case dns.TypeA:
+	case dns.TypeAAAA:
+	case dns.TypeANY:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question type: %v", dns.Type(question.Qtype).String())
+	}
+
+	serviceInstances, err := s.retrieveServices(question, request, response)
+
+	if err != nil {
+		return err
+	}
+
+	err = s.findMatchingServices(question, request, response, serviceInstances)
+	return err
+
+}
+
+func (s *Server) retrieveServices(question dns.Question, request, response *dns.Msg) ([]*client.ServiceInstance, error) {
+	// parse query :
+	// Query format:
+	// [tag]*.<service>.<domain>.
+	numberOfLabels, isValidDomain := dns.IsDomainName(question.Name)
+	if !isValidDomain {
+		response.SetRcode(request, dns.RcodeBadName)
+		return nil, fmt.Errorf("Invalid Domain name %s", question.Name)
+	}
+
+	fullDomainRequestArray := dns.SplitDomainName(question.Name)
+	if len(fullDomainRequestArray) == 1 {
+		response.SetRcode(request, dns.RcodeNameError)
+		return nil, fmt.Errorf("service name wasn't included in domain %s", question.Name)
+
+	}
+	serviceName := fullDomainRequestArray[numberOfLabels-2]
+	tags := fullDomainRequestArray[:len(fullDomainRequestArray)-2]
+	var ServiceInstances []*client.ServiceInstance
+	var err error
+	if len(tags) == 0 {
+		ServiceInstances, err = s.config.DiscoveryClient.ListServiceInstances(serviceName)
+	} else {
+		filters := client.InstanceFilter{ServiceName: serviceName, Tags: tags}
+		ServiceInstances, err = s.config.DiscoveryClient.ListInstances(filters)
+	}
+	if err != nil {
+		// TODO: what Error should we return ?
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return nil, fmt.Errorf("Error while reading from registry: %s", err.Error())
+	}
+	return ServiceInstances, nil
+}
+
+func (s *Server) findMatchingServices(question dns.Question, request, response *dns.Msg,
+	serviceInstances []*client.ServiceInstance) error {
+	numOfMatchingRecords := 0
+	for _, serviceInstance := range serviceInstances {
+		endPointType := serviceInstance.Endpoint.Type
+		var ip net.IP
+		var err error
+
+		switch endPointType {
+		case "tcp":
+			ip, err = validateEndPointTypeTCPAndUDP(serviceInstance.Endpoint.Value)
+
+		case "udp":
+			ip, err = validateEndPointTypeTCPAndUDP(serviceInstance.Endpoint.Value)
+
+		case "http":
+			ip, err = validateEndPointTypeHTTP(serviceInstance.Endpoint.Value)
+
+		default:
+			continue
+		}
+		if err != nil {
+			// TODO: Do I need to skip or should I print something ?
+			continue
+		}
+		numOfMatchingRecords++
+		if ip.To4() != nil {
```

Comment:
You can use the dns.QuestionType here to determine if the user wants a A, AAAA or a SRV record and return the appropriate list.